global.go

```diff
@@ -0,0 +1,248 @@
+// function. A maximization problem can be transformed into a
+// minimization problem by multiplying the function by -1.
+//
+// Global, unlike Local, searches for the global minimum of the function.
+// GlobalMethods typically do not make any assumptions about the number and location
+// of local minima, and so the only convergence metric used is the convergence
+// of function values found at major iterations. GradientThreshold is not checked
+// during convergence, and initial values are not used.
+func Global(p Problem, dim int, settings *Settings, method GlobalMethod) (*Result, error) {
+	if method == nil {
+		// TODO(btracey): Replace with a better global minimizer (basically
+		// anything else). Though technically this isn't worse on the space of
+		// all problems.
+		method = &GuessAndCheck{}
+	}
+	if settings == nil {
+		settings = DefaultSettingsGlobal()
+	}
+	startTime, err := initializeOptimization(p, settings, method)
+	if err != nil {
+		return nil, err
+	}
+	optLoc := newLocation(dim, method)
+	optLoc.F = math.Inf(1)
+
+	gs := &globalStatus{
+		mux:       &sync.Mutex{},
+		stats:     &Stats{},
+		status:    NotTerminated,
+		p:         p,
+		startTime: startTime,
+		optLoc:    optLoc,
+		settings:  settings,
+	}
+
+	nTasks := settings.Concurrent
+	method.InitGlobal(nTasks)
+
+	// Launch optimization workers
+	var wg sync.WaitGroup
+	for task := 0; task < nTasks; task++ {
+		wg.Add(1)
+		go func(task int) {
+			defer wg.Done()
+			loc := newLocation(dim, method)
+			x := make([]float64, dim)
+			globalWorker(task, method, gs, loc, x)
+		}(task)
+	}
+	wg.Wait()
+
+	// Cleanup and collect results
+	if settings.Recorder != nil && err == nil {
+		// Send the optimal location to Recorder.
+		err = settings.Recorder.Record(gs.optLoc, PostIteration, gs.stats)
+	}
+	gs.stats.Runtime = time.Since(startTime)
+	return &Result{
+		Location: *gs.optLoc,
+		Stats:    *gs.stats,
+		Status:   gs.status,
+	}, err
+}
+
+type globalStatus struct {
+	mux       *sync.Mutex
+	stats     *Stats
+	status    Status
+	p         Problem
+	startTime time.Time
+	optLoc    *Location
+	settings  *Settings
+}
+
+func globalWorker(task int, m GlobalMethod, g *globalStatus, loc *Location, x []float64) {
+	for {
+		// Find Evaluation location
+		op, err := m.IterateGlobal(task, loc)
+		if err != nil {
+			// TODO(btracey): Figure out how to handle errors properly. Shut
+			// everything down? Pass to globalStatus so it can shut everything down?
+			panic(err)
+		}
+
+		// Evaluate location and/or update stats.
+		status := g.globalOperation(op, loc, x)
+		if status != NotTerminated {
+			break
+		}
+	}
+}
+
+// globalOperation updates handles the status received by an individual worker.
+// It uses a mutex to protect updates where necessary.
+func (g *globalStatus) globalOperation(op Operation, loc *Location, x []float64) Status {
+	// TODO(btracey): Combine this with Local
+	stats := g.stats
+	mux := g.mux
```

Comment:
Agreed. The original code had mux as a function input, and so this was the easiest change to get something to compile. This way seemed slightly easier to get the flavor of what's going on, so I left it. This will definitely be changed.